applications/ParticleMechanicsApplication/custom_constitutive/johnson_cook_thermal_plastic_3D_law.cpp

```diff
@@ -0,0 +1,476 @@
+//    |  /           |
+//    ' /   __| _` | __|  _ \   __|
+//    . \  |   (   | |   (   |\__ `
+//   _|\_\_|  \__,_|\__|\___/ ____/
+//                   Multi-Physics
+//
+//  License:		BSD License
+//					Kratos default license: kratos/license.txt
+//
+//  Main authors:    Peter Wilson
+//
+
+// System includes
+
+// External includes
+
+// Project includes
+
+#include "custom_constitutive/johnson_cook_thermal_plastic_3D_law.hpp"
+#include "particle_mechanics_application_variables.h"
+
+namespace Kratos
+{
+	JohnsonCookThermalPlastic3DLaw::JohnsonCookThermalPlastic3DLaw()
+		: HyperElastic3DLaw()
+	{  }
+
+
+	JohnsonCookThermalPlastic3DLaw::JohnsonCookThermalPlastic3DLaw(const JohnsonCookThermalPlastic3DLaw& rOther)
+		: HyperElastic3DLaw(rOther)
+	{  }
+
+
+	ConstitutiveLaw::Pointer JohnsonCookThermalPlastic3DLaw::Clone() const
+	{
+		return Kratos::make_shared<JohnsonCookThermalPlastic3DLaw>(*this);
+	}
+
+
+	JohnsonCookThermalPlastic3DLaw::~JohnsonCookThermalPlastic3DLaw()
+	{  }
+
+
+	bool JohnsonCookThermalPlastic3DLaw::Has(const Variable<double>& rThisVariable)
+	{
+		if (rThisVariable == MP_TEMPERATURE
+			|| rThisVariable == MP_EQUIVALENT_PLASTIC_STRAIN
+			|| rThisVariable == MP_EQUIVALENT_PLASTIC_STRAIN_RATE
+			|| rThisVariable == MP_HARDENING_RATIO
+			|| rThisVariable == MP_EQUIVALENT_STRESS)
+			return true;
+		else return false;
+	}
+
+
+	void JohnsonCookThermalPlastic3DLaw::InitializeMaterial(const Properties& rMaterialProperties, const GeometryType& rElementGeometry, const Vector& rShapeFunctionsValues)
+	{
+		BaseType::InitializeMaterial(rMaterialProperties, rElementGeometry, rShapeFunctionsValues);
+
+		mStrainOld = ZeroVector(GetStrainSize());
+		mEquivalentPlasticStrainOld = 0.0;
+		mPlasticStrainRateOld = 0.0;
+		mEnergyInternal = 0.0;
+		mEnergyDissipated = 0.0;
+		mTemperatureOld = rMaterialProperties[TEMPERATURE];
+		mGammaOld = 1e-8;
+		mHardeningRatio = 1.0;
+
+		if (rMaterialProperties[TAYLOR_QUINNEY_COEFFICIENT] == 0.0) {
+			KRATOS_WARNING("Johnson Cook Material Model") << " Taylor Quinney Coefficient set to 0, ignoring thermal effects" << std::endl;
+		}
+
+		mYieldStressOld = CalculateHardenedYieldStress(rMaterialProperties, mEquivalentPlasticStrainOld, mPlasticStrainRateOld, mTemperatureOld);
+		mYieldStressVirgin = mYieldStressOld;
+	}
+
+
+	void JohnsonCookThermalPlastic3DLaw::CalculateMaterialResponseKirchhoff(Kratos::ConstitutiveLaw::Parameters& rValues)
```

Comment:
Normally we separate this to flow_rule, hardening_law, and yield_surface, see the MCC and MC laws. Do you think we could do the same here?

Reply:
This Material law is designed and tested for the specific components. Generally, it is possible but makes everything more complex which is not needed, here. No one will ever need all of those yield surfaces and all combination options.

Reply:
I do agree, I was suggesting this for extensibility and generalization, since the other laws in particle were written in that fashion.